[feat]: [IDP-306]: Sync merged app-config

sarthakkasat opened the pull request:
## Describe your changes

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger utAll`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions


## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/harness/harness-core/46291)
<!-- Reviewable:end -->

sarthakkasat commented:
trigger ut0

sandeepa-kv-jois commented on idp-service/contracts/openapi/v1/openapi.yaml:
```diff
@@ -631,6 +631,26 @@ paths:
       parameters:
         - $ref: '#/components/parameters/AccountHeader'
       x-internal: true
+  /v1/merged-app-config:
```
nit: update it to reflect what it's doing something like /v1/sync-merged-app-config or /v1/sync-app-config

sandeepa-kv-jois commented on idp-service/src/main/resources/baseappconfig.yaml:
```diff
@@ -1,5 +1,3 @@
   '/harness/prod':
-    target: 'https://app.harness.io/'
-    headers:
-      'x-api-key': ${HARNESS_QA_API_KEY}
```
Why is this x-api-key removed?, is it due to we are reading user session now from the browser localstorage?

sandeepa-kv-jois commented on idp-service/src/main/resources/app-config.yml:
```diff
@@ -0,0 +1,76 @@
+  auth:
+    keys:
+      - secret: ${BACKEND_SECRET}
+  baseUrl: https://${BACKEND_BASE_URL}/${NAMESPACE}/idp
```
- Please add a support for reading values for variables ${APP_BASE_URL}, ${BACKEND_BASE_URL},${POSTGRES_HOST} from ng-pr-manifest and ng-prod-manifest files.
- {NAMESPACE}, {ACCOUNT_ID}: should be replaced in idp service before syncing appconfis

vikyathharekal commented on idp-service/src/main/java/io/harness/idp/configmanager/service/ConfigManagerServiceImpl.java:
```diff
@@ -295,4 +303,15 @@ private List<String> getAllEnvVariablesForMultiplePluginIds(String accountIdenti
   }
+
+  @Override
+  public void checkUserAuthorization() {
```
There's already a method for this in ProvisionServiceImpl. Can we try using that here?

vikyathharekal replied:
There's already a method for this in ProvisionServiceImpl. Can we try using that here?

devesh-harness reviewed (changes requested):
> Reviewed 2 of 7 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/46291)* status: 7 file(s) to review. 3 change requests. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`idp-service/src/main/java/io/harness/idp/configmanager/resource/MergedPluginsConfigApiImpl.java` line 33 at r2](https://reviewable.io/reviews/harness/harness-core/46291#-NSo2Upd106lrwXJWWtN:-NSo2Upd106lrwXJWWtO:b-21m1mt) ([raw file](https://github.com/harness/harness-core/blob/9e043be6ed0a85146b2e69884fc2b23a8f3ca0c2/idp-service/src/main/java/io/harness/idp/configmanager/resource/MergedPluginsConfigApiImpl.java#L33)):*
> ```Java
>   private static final String APP_CONFIG_PATH = "app-config.yml";
>   private static final String BASE_CONFIG_NAME = "backstage-base-config";
>   private static final String CONFIG_NAME = "backstage-override-config";
> ```

lets rename the variable name to OVERRIDE_CONFIG_NAME
___
*[`idp-service/src/main/resources/baseappconfig.yaml` line 0 at r2](https://reviewable.io/reviews/harness/harness-core/46291#gh-1163462963:-NSo46t3CwHPMjR7FBag:b-t7yqdy) ([raw file](https://github.com/harness/harness-core/blob/9e043be6ed0a85146b2e69884fc2b23a8f3ca0c2/idp-service/src/main/resources/baseappconfig.yaml#L0)):*
<details><summary><i>Previously, sandeepa-kv-jois wrote…</i></summary><blockquote>

*(Reviewable was [unable](https://github.com/Reviewable/Reviewable/issues/14) to map this GitHub inline comment thread to the right spot — sorry!)*

Why is this x-api-key removed?, is it due to we are reading user session now from the browser localstorage?
</blockquote></details>

yes this is not needed now, check with Deba.


<!-- Sent from Reviewable.io -->

devesh-harness reviewed (commented):
> Reviewed 3 of 7 files at r1.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/46291)* status: 7 file(s) to review. 3 change requests. 1. unresolved discussions. 0 of 2 approvals obtained. 



<!-- Sent from Reviewable.io -->

sarthakkasat commented:
*[`idp-service/src/main/java/io/harness/idp/configmanager/resource/MergedPluginsConfigApiImpl.java` line 33 at r2](https://reviewable.io/reviews/harness/harness-core/46291#-NSo2Upd106lrwXJWWtN:-NSqid1P1MSpG9Q7h2x-:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/9e043be6ed0a85146b2e69884fc2b23a8f3ca0c2/idp-service/src/main/java/io/harness/idp/configmanager/resource/MergedPluginsConfigApiImpl.java#L33)):*
<details><summary><i>Previously, devesh-harness wrote…</i></summary><blockquote>

lets rename the variable name to OVERRIDE_CONFIG_NAME
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

sarthakkasat commented:
*[`idp-service/src/main/java/io/harness/idp/configmanager/service/ConfigManagerServiceImpl.java` line 308 at r2](https://reviewable.io/reviews/harness/harness-core/46291#gh-1163564211:-NSqig738V52TjcznNUY:bpiul8n) ([raw file](https://github.com/harness/harness-core/blob/9e043be6ed0a85146b2e69884fc2b23a8f3ca0c2/idp-service/src/main/java/io/harness/idp/configmanager/service/ConfigManagerServiceImpl.java#L308)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

There's already a method for this in ProvisionServiceImpl. Can we try using that here?
</blockquote></details>

Moved the method to commonUtils and updated it in provisionServiceApi too 


<!-- Sent from Reviewable.io -->
